src/axl.c

```diff
@@ -57,14 +58,84 @@ Global Variables
  * before transferring files */
 int axl_make_directories;
 
-/* AXL's flush file, SCR has one as well */
-char* axl_flush_file = NULL;
+/*
+ * Array for all the AXL_Create'd kvtree pointers.  It's indexed by the AXL id.
+ *
+ * Note: We only expand this array, we never shrink it.  This is fine since
+ * the user is only going to be calling AXL_Create() a handful of times.  It
+ * also simplifies the code if we never shrink it, and the extra memory usage
+ * is negligible, if any at all.
+ */
+kvtree** axl_kvtrees;
+static unsigned int axl_kvtrees_count = 0;
+
+static pthread_mutex_t id_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/*
+ * Allocate a new kvtree and return the AXL ID for it.  If state_file is
+ * specified, then populate the kvtree with it's data.
+ */
+static int axl_alloc_id(const char* state_file)
+{
+    kvtree* new;
+
+    new = kvtree_new();
+    if (state_file) {
+         /* initialize kvtree values from state_file if we have one */
+        if (access(state_file, F_OK) == 0 &&
+            kvtree_read_file(state_file, new) != KVTREE_SUCCESS) {
+            AXL_ERR("Couldn't read state file correctly");
+            return -1;
+        }
+
+        kvtree_util_set_str(new, AXL_KEY_STATE_FILE, state_file);
+    }
+
+    pthread_mutex_lock(&id_lock);
+    axl_kvtrees = realloc(axl_kvtrees, sizeof(struct kvtree *) * (axl_kvtrees_count + 1));
+    axl_kvtrees[axl_kvtrees_count] = new;
+    axl_kvtrees_count++;
+    pthread_mutex_unlock(&id_lock);
+    return axl_kvtrees_count - 1;
+}
+
+/* Remove the state file for an id, if one exists */
+static void axl_remove_state_file(int id)
+{
+    kvtree* file_list = axl_kvtrees[id];
+    char *state_file = NULL;
+
+    if (kvtree_util_get_str(file_list, AXL_KEY_STATE_FILE,
+        &state_file) == KVTREE_SUCCESS) {
+        axl_file_unlink(state_file);
+    }
+}
 
-/* Transfer handle unique IDs */
-static int axl_next_handle_UID = -1;
+/* Remove the ID, we don't need it anymore */
+static void axl_free_id(int id)
+{
+    pthread_mutex_lock(&id_lock);
+
+    /* kvtree_delete() will set axl_kvtrees[id] = NULL */
+    kvtree_delete(&axl_kvtrees[id]);
```

Comment:
is the tree being deleted here, and then a few lines later accessed in the call to axl_remove_state_file?

Reply:
Good catch.  I'll fix that